shared/subspace-data-retrieval/src/segment_fetcher.rs

```diff
@@ -85,66 +80,87 @@ where
-                            "Piece request failed",
-                        );
-                        return None;
-                    }
-                };
-
-                trace!(?piece_index, "Piece request succeeded");
-
-                // Piece was received successfully, "remove" this slot from semaphore
-                permit.forget();
-                Some((piece_index, piece))
-            }
-        })
-        .collect::<FuturesUnordered<_>>();
-
+    // We want NUM_RAW_RECORDS pieces to reconstruct the segment, but it doesn't matter exactly which ones.
+    let piece_indexes = segment_index.segment_piece_indexes_source_first();
+    let mut piece_indexes = piece_indexes.as_slice();
     let mut segment_pieces = vec![None::<Piece>; ArchivedHistorySegment::NUM_PIECES];
+
+    let mut pieces_pending = 0;
     let mut pieces_received = 0;
+    let mut piece_streams = Vec::new();
+
+    // Loop Invariant:
+    // - the number of remaining piece indexes gets smaller, eventually finishing the fetcher, or
+    // - the number of pending pieces gets smaller, eventually triggering another batch.
+    // We also exit early if we have enough pieces to reconstruct a segment.
+    'fetcher: while !piece_indexes.is_empty()
+        && pieces_received < RecordedHistorySegment::NUM_RAW_RECORDS
+    {
+        // Request the number of pieces needed to reconstruct the segment, assuming all pending
+        // pieces are successful.
+        let batch_size = RecordedHistorySegment::NUM_RAW_RECORDS - pieces_received - pieces_pending;
+        if batch_size > 0 {
+            let (batch_indexes, remaining_piece_indexes) = piece_indexes
+                .split_at_checked(batch_size)
+                .unwrap_or((piece_indexes, &[]));
+            // Invariant: the number of remaining piece indexes gets smaller.
+            piece_indexes = remaining_piece_indexes;
+
+            let stream = piece_getter.get_pieces(batch_indexes.iter().cloned()).await;
+            match stream {
+                Ok(stream) => {
+                    piece_streams.push(stream);
+                    pieces_pending += batch_size;
+                }
+                Err(error) => {
+                    // A single batch failure isn't fatal.
+                    debug!(
+                        ?error,
+                        %segment_index,
+                        batch_size,
+                        pieces_pending,
+                        pieces_received,
+                        pieces_needed = RecordedHistorySegment::NUM_RAW_RECORDS,
+                        "Segment piece getter batch failed"
+                    );
+                }
+            }
+        }
 
-    while let Some(maybe_piece) = received_segment_pieces.next().await {
-        let Some((piece_index, piece)) = maybe_piece else {
-            continue;
+        // Poll all the batches concurrently, getting all finished pieces.
+        let piece_responses = stream::iter(&mut piece_streams)
```

Comment:
Note to self: `Vec::retain(!Fuse::is_done)` before polling, to drop finished streams.